README.md
# PHP SDK for the Weather Source API Version 3.1


## Requirements

Requires PHP version 5.3.0 or greater.


## Installation

1. Save api.weathersource.sdk directory to your application
2. Rename api.weathersource.sdk/sdk/config_sample.php to api.weathersource.sdk/sdk/config.php
3. Update api.weathersource.sdk/sdk/config.php with your credentials and preferences
4. Include api.weathersource.sdk/sdk/weathersource_api.sdk.php in your application
5. Use the SDK at will

To acquire valid credentials, sign up for an API account directly at <http://weathersource.com>


Weather_Source_API_Requests class
---------------------------------

Include the class definition:

    include_once('/path/to/api.weathersource.sdk/sdk/weather_source_api_requests.php');

With the proper options set in api.weathersource.sdk/sdk/config.php, this class can:

1. Retry requests N times until success
2. Log errors to specified directory
3. Convert imperial measurement results to metric values
4. Convert Fahrenheit results to Celsius values
5. Suppress HTTP error codes
6. Return diagnostic information with responses

## Use

The Weather_Source_API_Requests class (alias: Weather_Source_API_Request) is designed for multithreaded requests. Add multiple requests, and they will be processed as thread capacity allows.

You can optionally define a callback function that will process results as they come available. This will allow processing of results while unprocessed requests continue to complete.

    /**
     *
     *  User defined callback function to process results as individual requests complete
     *
     *  @param   $result     string  The response to the cURL response. Prepend with '&' to pass by reference.
     *  @param   $http_code  string  The HTTP code generated by the cURL request
     *  @param   $latency    float   Seconds elapsed since node added accurate to the nearest microsecond
     *  @param   $url        string  User provided URL
     *  @param   $opts       array   The cURL transfer options
     *
     *  @return  NULL
     *
     */
    function user_defined_callback( &$result, $http_code, $latency, $url, $opts ) {

        // do something to $result
    }


Make as many requests as you like. Please reference the API documentation at <http://developer.weathersource.com> for valid request methods, request paths, and request parameters:

    $request_1 = new Weather_Source_API_Request(
        $request_method     = 'GET',
        $request_path       = 'account',
        $request_parameters = array( 'fields' => 'key,username' ),
        $callback_name      = 'user_defined_callback'  // optional
    );
    $request_2 = new Weather_Source_API_Request(
        $request_method     = 'GET',
        $request_path       = 'account',
        $request_parameters = array( 'fields' => 'first_name,last_name,email' ),
        $callback_name      = 'user_defined_callback'  // optional
    );

Using two static methods, we will now wait for the threaded requests to complete, and grab the results of all requests made:

    Weather_Source_API_Requests::finish();
    $results = Weather_Source_API_Requests::get_results();

There are also two instance methods at our disposal:

    $status_1 = $request_1->get_status();  // returns 'queued', 'processing', 'complete', or 'unknown'
    $result_1 = $request_1->get_result();  // returns an array containing the request response if the request status
                                           // is 'complete', FALSE otherwise



change_log.md
Change Log
==========

Version 3.1
-----------
Code Cleanup & Formatting
Removed trailing ?>'s

Version 3.0
-----------

Fixed bug that prevented request retries from operating correctly.
Added capability to enable verbose output and/or progress output.
Added debug output capability.


Version 2.4
-----------

Removed WSSDK_REQUEST_RETRY_ON_ERROR_DELAY from config_sample.php and set the retry delay
    to double the expected delay per retry attempt.


Version 2.3
-----------

Fixed bug where sleep was for seconds rather than microseconds.


Version 2.2
-----------

Added cURL option to prevent HTTP 100 responses
Improved memory efficiency
Added warm-up scaling logic to retry requests


Version 2.1
-----------

Added warm-up scaling feature to reduce connection errors when making high volumes of
	requests. By default, an initial 1000 requests per minute are allowed, doubling
	every 7 minutes until the Max Requests Per Minute is reached.


Version 2.0
-----------

Rearchitected for multi-threaded requests.
Changes to config_sample.php are not backwards compatable. Update your config.php file.
Code written for Version 1.X is not compatable with 2.0, and must be ported.

demo/index.php
<?php
/**
 * @author Jeffrey D. King
 * @copyright 2012-2015 Weather Source, LLC
 * @since Version 3.1
 */


/*  initiate our API class instance  */

require_once( __DIR__ . '/../sdk/weather_source_api_requests.php' );


/*  define a callback function to handle requests as they complete  */

/**
 *  User defined callback function to process results as individual requests complete
 *
 *  @param   $result     string  The response to the cURL response. Prepend with '&' to pass by reference.
 *  @param   $http_code  string  The HTTP code generated by the cURL request
 *  @param   $latency    float   Seconds elapsed since node added accurate to the nearest microsecond
 *  @param   $url        string  User provided URL
 *  @param   $opts       array   The cURL transfer options
 *  @return  NULL
 */
function user_defined_callback( &$result, $http_code, $latency, $url, $opts ){

    echo "<pre>callback \$result = ";
    print_r( $result );
    echo "</pre>";

    // unset the key field. Since $result was passed by reference, this will change the final $results
    unset($result['key']);

} // End of function user_defined_callback()


/*  make multiple API requests  */

for( $i=0; $i<10; $i++ ){


    /*  make an API request  */

    $request = new Weather_Source_API_Request(
        $request_method     = 'GET',
        $request_path       = 'account',
        $request_parameters = array( 'fields' => 'key,username,first_name,last_name,email' ),
        $callback_name      = 'user_defined_callback' // optional
    );

    // $request->get_status() will return 'queued', 'processing', 'complete', or 'unknown'
    echo "request status = {$request->get_status()}<br>";

} // End of for()


/*  wait for all requests to complete  */

Weather_Source_API_Requests::finish();


/*  get the results ad do something with them  */

$results = Weather_Source_API_Requests::get_results();

echo "<pre>summary \$results = ";
print_r( $results );
echo "\n</pre>";


/*  get the last submitted request's result  */

if( 'complete' == $request->get_status() ) {

    $result = $request->get_result();
    echo "<pre>last request \$result = ";
    print_r( $result );
    echo "</pre>";

} // End of if()


sdk/config_sample.php
<?php
/**
 * @author Jeffrey D. King
 * @copyright 2012-2015 Weather Source, LLC
 * @since Version 3.1
 */

//==============================================================================
//==    IMPORTANT:	after setting the configuration options below,
//==    			rename this file to "config.php"
//==============================================================================


/**
 *	@const  string  [REQUIRED]  Your API key
 */
define('WSAPI_KEY', '' );

/**
 *	@const  string  [OPTIONAL]  The base API URI. If not defined here, WSAPI_BASE_URI defaults to
 *							    'https://api.weathersource.com'.
 */
define('WSAPI_BASE_URI', 'https://api.weathersource.com' );

/**
 *	@const  string  [OPTIONAL]  The API version. Make sure the version number is prepended with a
 *							    'v'. If not defined here, WSAPI_VERSION defaults to 'v1'.
 */
define('WSAPI_VERSION', 'v1'  );

/**
 *	@const  boolean  [OPTIONAL]  Return diagnostic information with response? If not defined here,
 *							     WSAPI_RETURN_DIAGNOSTICS defaults to FALSE.
 */
define('WSAPI_RETURN_DIAGNOSTICS', FALSE );

/**
 *	@const  boolean  [OPTIONAL]  Suppress all HTTP response codes (i.e. force a 200 response)? If
 *							     not defined here, WSAPI_SUPPRESS_RESPONSE_CODES defaults to FALSE.
 */
define('WSAPI_SUPPRESS_RESPONSE_CODES', FALSE );

/**
 *	@const  integer  [OPTIONAL]  Pace requests to comply with a subscription plan's minute rate limit.
 *							     If not defined here, WSSDK_MAX_REQUESTS_PER_MINUTE defaults to 10.
 */
define('WSSDK_MAX_REQUESTS_PER_MINUTE', 10 );

/**
 *	@const  integer  [OPTIONAL]  The maximum thread count. WARNING: To many threads originating from
 *								 the same IP address may result in connection errors! If not defined
 *							     here, WSSDK_MAX_THREADS defaults to 10.
 */
define('WSSDK_MAX_THREADS', 10 );

/**
 *	@const  string  [OPTIONAL]  Defines the unit type to return for relevant observations. Allowed
 *								values are 'imperial' and 'metric'. If set to imperial, relevant
 *								values will be returned as inches or miles per hour. If set to
 *								metric, relevant values will be returned as centimeters or
 *								kilometers per hour. This will only convert responses from the API.
 *								All writes must conform to the API spec (imperial). If not defined
 *							    here, WSSDK_DISTANCE_UNIT defaults to 'imperial'.
 */
define('WSSDK_DISTANCE_UNIT', 'imperial' );

/**
 *	@const  string  [OPTIONAL]  Defines the unit type to return for relevant observations. Allowed
 *								values are 'fahrenheit' and 'celsius'. This will only convert responses
 *								from the API. All writes must conform to the API spec (fahrenheit).
 *							    If not defined here, WSSDK_TEMPERATURE_UNIT defaults to 'fahrenheit'.
 */
define('WSSDK_TEMPERATURE_UNIT', 'fahrenheit' );

/**
 *	@const  boolean  [OPTIONAL]  Should all errors be written to a log file? If WSAPI_SUPPRESS_RESPONSE_CODES
 *								 is set to TRUE, this option will be forced to FALSE. If not defined here,
 *							     WSSDK_LOG_ERRORS defaults to FALSE.
 */
define('WSSDK_LOG_ERRORS', FALSE );

/**
 *	@const  string  [OPTIONAL]  Log file directory location. A path beginning beginning with '/' is
 *								considered absolute, otherwise, it is treated as relative to the
 *								api.weathersource.sdk/sdk directory. If not defined here,
 *							    WSSDK_ERROR_LOG_DIRECTORY defaults to 'error_logs/'.
 */
define('WSSDK_ERROR_LOG_DIRECTORY', 'error_logs/' );

/**
 *	@const  integer  [OPTIONAL]  The number of times to retry a request that returns a potentially
 *								 recoverable error. Retries are exponentially delayed. If not defined
 *								 here, WSSDK_REQUEST_RETRY_ON_ERROR_COUNT defaults to 10.
 */
define('WSSDK_REQUEST_RETRY_ON_ERROR_COUNT', 10 );

/**
 *	@const  integer  [OPTIONAL]  The initial allowable requests per minute for warm-up scaling.
 *							     WARNING: increasing this number may result in connection errors
 *							     while the API infrustructure scales to meet your high demand.
 *							     If not defined here, WSSDK_SCALING_INITIAL_REQUESTS_PER_MINUTE
 *							     defaults to 1000.
 */
define('WSSDK_SCALING_INITIAL_REQUESTS_PER_MINUTE', 1000 );

/**
 *	@const  float  [OPTIONAL]  The number of minutes to double warm-up scaling allowable requests
 *							   per minute. WARNING: decreasing this number may result in connection
 *							   errors while the API infrustructure scales to meet your high demand.
 *							   If not defined here, WSSDK_SCALING_DOUBLE_CAPACITY_MINUTES defaults
 *							   to 7.
 */
define('WSSDK_SCALING_DOUBLE_CAPACITY_MINUTES', 7 );


sdk/curl_node.php
<?php
/**
 * @author     Jeffrey D. King
 * @copyright  2012-2015 Weather Source, LLC
 * @since      Version 3.1
 */

/**
 *  Class to manage multithreaded cURL requests
 *
 *  Example usage:
 *
 *  <code>
 *
 *      // Process results as individual requests complete (rather than waiting
 *      // for all requests to complete) by defining a callback function.
 *
 *      /**
 *       *  User defined callback function to process results as the individual request completes
 *       *
 *       *  @ param   $response   string  The response to the cURL response. Prepend with '&' to pass by reference.
 *       *  @ param   $metadata   string  User defined metadata associated with the request
 *       *  @ param   $http_code  string  The HTTP code generated by the cURL request
 *       *  @ param   $latency    float   Seconds elapsed since node added accurate to the nearest microsecond
 *       *  @ param   $url        string  User provided URL
 *       *  @ param   $opts       array   The cURL transfer options
 *       *  @ return  whatever you define
 *       * /
 *      function user_defined_callback( &$response, $metadata, $http_code, $latency, $url, $opts ) {
 *
 *          // do something with $response
 *      }
 *
 *      // URLs to requests
 *      $urls = array(
 *          'http://www.example.com',
 *          'http://www1.example.com',
 *          'http://www2.example.com',
 *      );
 *
 *      // Pace your requests:
 *      Curl_Node::set_max_requests_per_minute(60);
 *
 *      // Only allow 10 threads at a time
 *      Curl_Node::set_max_threads(10);
 *
 *      foreach( $urls as $url ) {
 *
 *          // Add a cURL node to be processed as thread availability allows
 *          $node->new Curl_Node(
 *              $url,
 *              array( CURLOPT_TIMEOUT => 60, CURLOPT_CONNECTTIMEOUT => 5 ),
 *              'user_defined_callback'
 *          );
 *
 *          // use of the two instance methods
 *          if( 'complete' == $node->get_status() ) {
 *              $result = $node->get_result();
 *          }
 *      }
 *
 *      // Once all the nodes are added, wait for requests to complete
 *      Curl_Node::finish();
 *
 *      // Get an array of results
 *      // $results format:
 *      //     array(
 *      //        array(
 *      //            'response'  => string,    // Response message
 *      //            'metadata'  => string,    // User defined metadata associated with the request
 *      //            'http_code' => mixed,     // HTTP response code, 0 for cURL error
 *      //            'latency'   => float,     // Seconds elapsed since node added accurate to the nearest microsecond
 *      //            'url'       => string,    // User provided URL
 *      //            'opts'      => array,     // User provided cURL transfer options
 *      //        ),
 *      //        ...
 *      //     ),
 *      $results = Curl_Node::get_results();
 *
 *  </code>
 */
class Curl_Node {

    /**
     *  @access  private
     *  @static
     *  @var     array  Stack of queued requests
     */
    static private $queue;

    /**
     *  @access  private
     *  @static
     *  @var     array  Stack of active requests
     */
    static private $active;

    /**
     *  @access  private
     *  @static
     *  @var     array  Stack of completed requests
     */
    static private $complete;

    /**
     *  @access  private
     *  @static
     *  @var     resource  Global store for the cURL multi handle
     */
    static private $multi_handle;

    /**
     *  @access  private
     *  @static
     *  @var     integer  Count of active threads
     */
    static private $threads;

    /**
     *  @access  private
     *  @static
     *  @var     integer  Status of the multi cURL stack
     */
    static private $status;

    /**
     *  @access  private
     *  @static
     *  @var     boolean  Indicates whether nodes are currently being added.
     *                    Used to lock self::add_nodes() to a single concurrent
     *                    loop.
     */
    static private $add_nodes_lock;

    /**
     *  @access  private
     *  @static
     *  @var     integer  Upper limit for quantity of threads
     */
    static private $max_threads = 1;

    /**
     *  @access  private
     *  @static
     *  @var     integer  The maximum number of retries if a request suffers
     *                    a potentially recoverable failure.
     */
    static private $max_retries = 10;

    /**
     *  @access  private
     *  @static
     *  @var     integer  Upper limit for the number of requests allowed
     *                    per minute
     */
    static private $max_requests_per_minute = 10;

    /**
     *  @access  private
     *  @static
     *  @var     integer  Initial allowable requests per minute for
     *                    warm-up scaling.
     */
    static private $scaling_initial_requests_per_minute = 1000;

    /**
     *  @access  private
     *  @static
     *  @var     float  The number of minutes to double warm-up scaling
     *                  allowable requests per minute.
     */
    static private $scaling_double_capacity_minutes = 7;

    /**
     *  @access  private
     *  @static
     *  @var     integer  The UNIX timestamp when warm-up scaling began.
     */
    static private $scaling_initialization_timestamp;

    /**
     *  @access  private
     *  @static
     *  @var     boolean  If set to true, every time a request is queued, '+' is output,
     *                    every time a request is made, '-' is output, and every time a
     *                    response is received for a request '×' is output.
     */
    static private $show_progress = FALSE;

    /**
     *  @access  private
     *  @static
     *  @var     boolean  Log debug information?
     */
    static private $debug = FALSE;

    /**
     *  @access  private
     *  @static
     *  @var     string  All debug information is stored to this property.
     */
    static private $debug_info = '';

    /**
     *  @access  private
     *  @var     string  An individual cURL handle cast to a string. Used as
     *                   a unique id for efficient lookup from the self::$queue,
     *                   self::$active, and self::$complete stacks
     */
    private $handle_string;

    /**
     *  Initiate a class instance and add a request to multithreaded cURL handler
     *
     *  @access  public
     *  @param   string  $url       [REQUIRED]  URL for the cURL request
     *  @param   array   $opts      [OPTIONAL]  An array of cURL options formatted like:
     *                                          array( CURLOPT_URL => 'http://example.com' )
     *  @param   callable   $callback  [OPTIONAL]  If provided, the user defined callback
     *                                          function will be called as this individual
     *                                          request completes. You do not need to wait
     *                                          for everything to finish!
     *  @param   array   $metadata  [OPTIONAL]  A container to attach any information you
     *                                          want to associate with this request.
     *  @return  NULL
     */
    public function __construct( $url, $opts = array(), $callback = '', $metadata = array() ){

        // set our warm-up scaling timestamp if it is not already set
        if( !isset(self::$scaling_initialization_timestamp) ){

            self::$scaling_initialization_timestamp = time();

        } // End of if()

        // force return of output
        $opts[CURLOPT_HEADER]         = 0;
        $opts[CURLOPT_RETURNTRANSFER] = 1;
        $opts[CURLOPT_URL]            = $url;
        $opts[CURLOPT_HTTPHEADER]     = array('Expect:'); // prevent HTTP 100:Continue responses

        $handle = curl_init();

        foreach($opts as $option => $value) {

            curl_setopt( $handle, $option, $value );

        } // End of foreach()

        $node['handle']       = $handle;
        $node['url']          = $url;
        $node['opts']         = $opts;
        $node['start']        = microtime(TRUE);
        $node['latency']      = 0;
        $node['retries']      = 0;
        $node['http_code']    = '';
        $node['response']     = '';
        $node['callback']     = $callback;
        $node['metadata']     = $metadata;

        self::$queue[(string) $handle] = $node;

        if( self::$debug ){

            self::$debug_info += ((string) $handle) . ": added to queue\n";

        } // End of if()

        if( self::$show_progress ){

            echo '+';

        } // End of if()

        self::request();

        $this->handle_string = (string) $handle;

    } // End of public function __construct()

    /**
     *  Set the initial allowable requests per minute for warm-up scaling.
     *
     *  @access  public
     *  @static
     *  @param   integer  $scaling_initial_requests_per_minute  [REQUIRED]  Allowable requests.
     *  @return  NULL
     */
    static public function set_scaling_initial_requests_per_minute( $scaling_initial_requests_per_minute ){

        self::$scaling_initial_requests_per_minute = $scaling_initial_requests_per_minute;

    } // End of static public function set_scaling_initial_requests_per_minute()


    /**
     *  Set the number of minutes to double warm-up scaling allowable requests per minute.
     *
     *  @access  public
     *  @static
     *  @param   float  $scaling_double_capacity_minutes  [REQUIRED]  Minutes.
     *  @return  NULL
     */
    static public function set_scaling_double_capacity_minutes( $scaling_double_capacity_minutes ){

        self::$scaling_double_capacity_minutes = $scaling_double_capacity_minutes;

    } // End of static public function set_scaling_double_capacity_minutes


    /**
     *  Set the maximum number of threads that may be utilized at any one time
     *
     *  @access  public
     *  @static
     *  @param   integer  $max_threads  [REQUIRED]  The maximum number of threads
     *  @return  NULL
     */
    static public function set_max_threads( $max_threads ){

        self::$max_threads = $max_threads;

    } // End of static public function set_max_threads()


    /**
     *  Set the number of seconds to wait between launching new threads to managably ramp up requests
     *
     *  @access  public
     *  @static
     *  @param   integer  $max_requests_per_minute  [REQUIRED]  The interval in seconds
     *                                                          between thread launches
     *  @return  NULL
     */
    static public function set_max_requests_per_minute( $max_requests_per_minute ){

        self::$max_requests_per_minute = $max_requests_per_minute;

    } // End of static public function set_max_requests_per_minute()


    /**
     *  Set debug logging
     *
     *  @access  public
     *  @static
     *  @param   boolean  $debug  [REQUIRED]  Log debug info?
     *  @return  NULL
     */
    static public function set_debug( $debug ){

        self::$debug = $debug;

    } // End of static public function set_debug()


    /**
     *  Set the maximum number of times to retry a recoverable error
     *
     *  @access  public
     *  @static
     *  @param   integer  $max_retries  [REQUIRED]  The maximum number of times to retry
     *                                              a recoverable error.
     *  @return  NULL
     */
    static public function set_max_retries( $max_retries ){

        self::$max_retries = $max_retries;

    } // End of static public function set_max_retries()


    /**
     *  Set self::$show_progress to TRUE or FALSE
     *
     *  @access  public
     *  @static
     *  @param   boolean  $show_progress  [REQUIRED]  If set to true, every time a request is queued,
     *                                                '+' is output, every time a request is made, '-'
     *                                                is output, and every time a response is received
     *                                                for a request '×' is output.
     *  @return  NULL
     */
    static public function show_progress( $show_progress ){

        self::$show_progress = $show_progress;

    } // End of static public function show_progress()


    /**
     *  Get the status of a node
     *
     *  @access  public
     *  @return  string   Possible values: "queued", "processing", "complete", "unknown"
     */
    public function get_status(){

        if( is_array(self::$complete) && isset(self::$complete[$this->handle_string]) ) {

            return 'complete';

        } elseif( is_array(self::$active) && isset(self::$active[$this->handle_string]) ) {

            return 'processing';

        } elseif( is_array(self::$queue) && isset(self::$queue[$this->handle_string]) ) {

            return 'queued';
        }

        return 'unknown';

    } // End of public function get_status()


    /**
     *  Get the result of a node request
     *
     *  @access  public
     *  @return  mixed   If a node request is not complete, returns FALSE.
     *                   If a node request has completed, returns an associative array
     *                   containing these keys: 'response' (string), 'http_code' (string),
     *                   'latency' (float), 'url' (string), 'opts' (array).
     */
    public function get_result(){

        if( isset(self::$complete[$this->handle_string]) ) {

            return self::$complete[$this->handle_string];

        } else {

            return FALSE;

        }

    } // End of public function get_result()


    /**
     *  Get all results. This method waits for all node requests to complete.
     *
     *  @access  public
     *  @static
     *  @return  array  An indexed array of associative arrays containing these keys:
     *                  'response' (string), 'http_code' (string), 'latency' (float),
     *                  'url' (string), 'opts' (array)
     */
    static public function get_results(){

        self::finish();

        return isset(self::$complete) ? self::$complete : array();

    } // End of static public function get_results()


    /**
     *  Get debug info.
     *
     *  @access  public
     *  @static
     *  @return  string  All debug information.
     */
    static public function get_debug_info(){

        return self::$debug_info;

    } // End of static public function get_debug_info()


    /**
     *  Wait for all request nodes to complete and process the results
     *
     *  @access  public
     *  @static
     *  @return  NULL
     */
    static public function finish(){

        if( self::$debug ){

            self::$debug_info += "Finishing\n";

        } // End of if()

        if( isset(self::$multi_handle) ){

            do {

                self::wait_for_result();
                self::process_results();

            } while( 0 < self::$threads && CURLM_OK == self::$status );

            curl_multi_close(self::$multi_handle);

            self::$multi_handle = NULL;

        } // End of if()

        if( self::$debug ){

            self::$debug_info += "Finished\n";

        } // End of if()

    } // End of static public function finish()



            /**
             *  Add a node and process any results that may have completed
             *
             *  @access  private
             *  @static
             *  @return  NULL
             */
            static private function request(){

                self::add_nodes();
                self::process_results();

            } // End of static private function request()



            /**
             *  Add a node
             *
             *  @access  private
             *  @static
             *  @return  NULL
             */
            static private function add_nodes(){

                if( 0 == count(self::$queue) ){

                    return;

                } // End of if()

                if( self::$add_nodes_lock !== TRUE ){

                    self::$add_nodes_lock = TRUE;
                    self::$multi_handle   = empty(self::$multi_handle) ? curl_multi_init() : self::$multi_handle;
                    self::$threads        = empty(self::$threads)      ? 0                 : self::$threads;
                    self::$status         = !isset(self::$status)      ? CURLM_OK          : self::$status;


                    /*  add nodes to curl requests until $max_threads added or $queue exhausted  */

                    while( self::$threads < self::$max_threads && 0 < count(self::$queue) && CURLM_OK == self::$status ){

                        // add node to $active
                        $node = array_shift(self::$queue);

                        self::$active[(string) $node['handle']] = $node;
                        curl_multi_add_handle( self::$multi_handle, $node['handle'] );

                        usleep( self::delay() * pow(2,$node['retries']) );

                        do {

                            self::$status = curl_multi_exec(self::$multi_handle, self::$threads);

                        } while( CURLM_CALL_MULTI_PERFORM == self::$status );

                        if( self::$debug ){

                            self::$debug_info += ((string) $node['handle']) . ": request submitted\n";

                        } // End of if()

                        if( self::$show_progress ){

                            echo '-';

                        } // End of if()

                    } // End of while()

                    self::$add_nodes_lock = FALSE;

                } // End of if()

            } // End of static private function add_nodes()


            /**
             *  Wait until activity on a node occurs
             *
             *  @access  private
             *  @static
             *  @return  NULL
             */
            static private function wait_for_result(){


                    if( CURLM_OK == self::$status && 0 < self::$threads && -1 != curl_multi_select(self::$multi_handle) ){

                        do {

                            if( self::$debug ){

                                self::$debug_info += "Waiting for response\n";

                            } // End of if()

                            self::$status = curl_multi_exec(self::$multi_handle, self::$threads);

                        } while (CURLM_CALL_MULTI_PERFORM == self::$status);

                    } // End of if()

            } // End of static private function wait_for_result()

            /**
             *  Process any results that may have completed
             *
             *  @access  private
             *  @static
             *  @return  NULL
             */
            static private function process_results(){

                $processed = FALSE;

                if( self::$debug ){

                    self::$debug_info += "Processing responses\n";

                } // End of if()

                while( FALSE !== ($info = curl_multi_info_read(self::$multi_handle)) ){

                    $processed = TRUE;
                    $handle    = $info['handle'];
                    $node      = self::$active[(string) $handle];
                    $http_code = curl_getinfo( $node['handle'], CURLINFO_HTTP_CODE );

                    unset(self::$active[(string) $handle]);

                    if( self::$debug ){

                        self::$debug_info += ((string) $handle) . ": processing response\n";

                    } // End of if()

                    if( in_array($http_code, array('0','403','408','500','503','504')) &&  $node['retries'] < self::$max_retries){

                        // we have an error that may be recovered from

                        $node['retries']++;

                        self::$queue[(string) $handle] = $node;

                        if( self::$debug ){

                            self::$debug_info += ((string) $handle) . ": recoverable error ({$http_code}), resubmitted to queue\n";

                        } // End of if()

                        if( self::$show_progress ){

                            echo " +{$node['retries']} ";

                        } // End of if()

                    } else {

                        $callback = $node['callback'];

                        $node['latency']    = microtime(TRUE) - $node['start'];
                        $node['http_code']  = (string) $http_code;
                        $node['response']   = ( $http_code != 0 ) ? curl_multi_getcontent($handle) : curl_error($handle);

                        unset($node['handle']);
                        unset($node['start']);
                        unset($node['retries']);
                        unset($node['callback']);

                        // we will for a $result array that allows $node['response'] to be passed by reference to user defined callback
                        $callback_params = array(
                            &$node['response'],
                            &$node['metadata'],
                            &$node['http_code'],
                            &$node['latency'],
                            &$node['url'],
                            &$node['opts']
                        );

                        if( !empty($callback) && is_callable($callback) ){

                            if( self::$debug ){

                                self::$debug_info += ((string) $handle) . ": sending response to callback function\n";

                            } // End of if()

                            call_user_func_array( $callback, $callback_params );

                        } // End of if()

                        // if the user set a node value to null in the callback function, just unset the parameter to save memory
                        if( $node['response']  === NULL ) unset($node['response']);
                        if( $node['metadata']  === NULL ) unset($node['metadata']);
                        if( $node['http_code'] === NULL ) unset($node['http_code']);
                        if( $node['latency']   === NULL ) unset($node['latency']);
                        if( $node['url']       === NULL ) unset($node['url']);
                        if( $node['opts']      === NULL ) unset($node['opts']);

                        self::$complete[(string) $handle] = $node;

                        if( self::$debug ){

                            self::$debug_info += ((string) $handle) . ": response processed\n";

                        } // End of if()

                        if( self::$show_progress ){

                            echo '×';

                        } // End of if()

                    } // End of if/else

                    curl_multi_remove_handle(self::$multi_handle, $handle);

                } // End of while()

                if( $processed && 0 < count(self::$queue) ){

                    self::add_nodes();

                } // End of if()

            } // End of static private function process_results()


            /**
             *  Get request delay in microseconds
             *
             *  @access  private
             *  @static
             *  @return  integer  The microseconds to delay between requests
             */
            static private function delay() {

                $scaling_minutes_elapsed     = (time() - self::$scaling_initialization_timestamp) / 60;
                $scaling_requests_per_minute = self::$scaling_initial_requests_per_minute * pow(2, $scaling_minutes_elapsed / self::$scaling_double_capacity_minutes);
                $scaling_delay_microseconds  = (60/$scaling_requests_per_minute) * 1000000;
                $min_delay_microseconds      = (60/self::$max_requests_per_minute) * 1000000;

                return max($scaling_delay_microseconds, $min_delay_microseconds);

            } // End of static private function delay()

} // End of class Curl_Node
